lib/plugindata/cas.go

```diff
@@ -0,0 +1,166 @@
+	GetPluginData(context.Context, types.PluginDataFilter) ([]types.PluginData, error)
+	UpdatePluginData(context.Context, types.PluginDataUpdateParams) error
+}
+
+// CompareAndSwap represents modifier struct
+type CompareAndSwap[T any] struct {
+	client      Client
+	name        string
+	kind        string
+	backoffBase time.Duration
+	backoffMax  time.Duration
+	encode      func(T) map[string]string
+	decode      func(map[string]string) T
+}
+
+// NewCAS returns modifier struct
+func NewCAS[T any](
+	client Client, name,
+	kind string,
+	encode func(T) map[string]string,
+	decode func(map[string]string) T,
+) *CompareAndSwap[T] {
+	return &CompareAndSwap[T]{
+		client,
+		name,
+		kind,
+		backoffBase,
+		backoffMax,
+		encode,
+		decode,
+	}
+}
+
+// Create tries to perform compare-and-swap update of a plugin data assuming that it does not exist
+//
+// fn callback function receives current plugin data value and returns modified value and
+// error.
+//
+// Please note that fn might be called several times due to CAS backoff, hence, you must be careful
+// with things like I/O ops and channels.
+func (c *CompareAndSwap[T]) Create(
+	ctx context.Context,
+	resource string,
+	newData T,
+) (T, error) {
+	emptyData := *new(T)
+
+	existingData, err := c.getPluginData(ctx, resource)
+	if err != nil && !trace.IsNotFound(err) {
+		return emptyData, trace.Wrap(err)
+	}
+
+	if existingData != nil {
+		return emptyData, trace.AlreadyExists("plugin data already exists")
+	}
+
+	err = c.updatePluginData(ctx, resource, newData, emptyData)
+	if err == nil {
+		return newData, nil
+	}
+
+	return emptyData, trace.Wrap(err)
+}
+
+// Update tries to perform compare-and-swap update of a plugin data assuming that it exist
+//
+// modifyT will receive existing plugin data and should return a modified version of the data.
+
+// If existing plugin data does not match expected data, then a trace.CompareFailed error should
+// be returned to backoff and try again.
+
+// To abort the update, modifyT should return an error other, than trace.CompareFailed, which
+// will be propagated back to the caller of `Update`.
+func (c *CompareAndSwap[T]) Update(
+	ctx context.Context,
+	resource string,
+	modifyT func(T) (T, error),
+) (T, error) {
+	emptyData := *new(T)
+
+	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
+	for {
+		oldData, err := c.getPluginData(ctx, resource)
+		if err != nil {
+			return emptyData, trace.Wrap(err)
+		}
+
+		cbData := *oldData
+		expectData := *oldData
+
+		newData, err := modifyT(cbData)
+		if trace.IsCompareFailed(err) {
+			err := backoff.Do(ctx)
+			if err != nil {
+				return emptyData, trace.Wrap(err)
+			}
+
+			return emptyData, trace.Wrap(err)
+		} else if err != nil {
+			return emptyData, trace.Wrap(err)
+		}
+
+		err = c.updatePluginData(ctx, resource, newData, expectData)
+		if err == nil {
+			return newData, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return emptyData, trace.Wrap(err)
+		}
```

Comment:
Don't we still need the backoff here as well?
```suggestion
		}
		if err := backoff.Do(ctx); err != nil {
			return emptyData, trace.Wrap(err)
		}
```